Chap_Terms.tex

```diff
@@ -4,39 +4,30 @@
+The following terms and terminology is used throughout this document:
 \begin{itemize}
-\item \declareterm{session}\emph{session} refers to an allocated set of resources assigned to a particular user by the system \ac{WLM}. Historically, \ac{HPC} sessions have consisted of a static allocation of resources - i.e., a block of resources are assigned to a user in response to a specific request and managed as a unified collection. However, this is changing in response to the growing use of dynamic programming models that require on-the-fly allocation and release of system resources. Accordingly, the term \emph{session} in this document refers to the current block of assigned resources and is a potentially dynamic entity.
-\item \declareterm{slot}\declareterm{slots}\emph{slot} refers to an allocated entry for a process. \acp{WLM} frequently allocate entire nodes to a \emph{session}, but can also be configured to define the maximum number of processes that can simultaneously be executed on each node. This often corresponds to the number of hardware \acp{PU} (typically cores, but can also be defined as hardware threads) on the node. However, the correlation between hardware \acp{PU} and slot allocations strictly depends upon system configuration.
-\item \declareterm{job}\emph{job} refers to a set of one or more \emph{applications} executed as a single invocation by the user within a session. For example, ``\textit{mpiexec -n 1 app1 : -n 2 app2}'' is considered a single \ac{MPMD} job containing two applications.
+\item \declareterm{session}\emph{session} refers to an allocated set of resources assigned to a particular user by the system \ac{WLM}. Historically, \ac{HPC} \emph{sessions} have consisted of a static allocation of resources - i.e., a block of resources are assigned to a user in response to a specific request and managed as a unified collection. However, this is changing in response to the growing use of dynamic programming models that require on-the-fly allocation and release of system resources. Accordingly, the term \emph{session} in this document refers to the current block of assigned resources and is a potentially dynamic entity.
+\item \declareterm{slot}\declareterm{slots}\emph{slot} refers to an allocated set of resources assigned to a process. \acp{WLM} frequently allocate entire nodes to a \emph{session}, but can also be configured to define the maximum number of processes that can simultaneously be executed on each node. This often corresponds to the number of hardware \acp{PU} (typically cores, but can also be defined as hardware threads) on the node. However, the correlation between hardware \acp{PU} and slot allocations strictly depends upon system configuration.
+\item \declareterm{application}\emph{application} refers to a single executable (binary, script, etc.) member of a \emph{job}. Applications consist of one or more \emph{processes}, either operating independently or cooperatively at any given time during their execution.
```

Comment:
What is a "single executable"?  What if you have multiple identical binaries copied across nodes? What about if you have a heterogeneous system where you need different versions for different ISAs from the same source? How does the use of accelerators, who have separate binaries for the different HW pieces, affect this statement? I think this is too imprecise.

Reply:
@schulzm I agree here too. Also, sometimes applications have multiple binaries acting in a loosely synchronous manner (e.g. a climate model with ocean and atmosphere binaries). Perhaps we can rephrase as:
"application refers to the executable(s) that comprise a job."

Reply:
I was thinking about this last night when preparing my talk for the ASC today. If we consider PMIx groups then it is possible that the only aspect that distinguishes the two groups of processes is the process set identifier.

For example (a modification of the example in Ch. 13.1):
```
prun -n 4 --pset ocean myapp -ocean : -n 3 --pset ice myapp -ice
```
There are two applications in this one job. They are distinct in their arguments, but not in their binaries. If we remove the arguments and assume that myapp will read it's process set to determine its mode of operation then the distinguishing aspect is the pset.

How would you all suggest we adjust the wording here to capture the full meaning of `application`?

Reply:
Aside but possibly relevant: Is the only way you can define a process set is by using PRRTE? How would you do it otherwise since there is no API for defining a process set? Does this launch the two different myapp as separate MPI_WORLDs?

Scanning through the document, it seems that what is meant by application is: "executables that run concurrently as part of a logical unit to compute a task." 

I think we can get away with saying "executables" and not specifying if they are the same or different, since they could be either and still be part of the same application.

Reply:
Why do we need to define this in the first place? Once a RM allocates the resources, it is up the user to decide what to do: running a set of scripts that subdivide the allocation, or one application binary looks the same to the RM, doesn't it? For the RM, what is the difference to a job or session?

Reply:
Per working group meeting Aug. 17, 2020
 - The term `application` is easily confused in HPC as the whole workflow (application = workflow) or a job (application = namd) or a grouping within the job (application = ocean subset of the job).
 - In PMIx we mean the latter (a subset of the job). Don't change the term, but let's clarify the definition.
 - Try this definition:
> A collection of processes that share the same expression of executable, environment, and other associated parameters within a job.
 - Hierarchy:
   * **Process** is smallest unit managed by PMIx
   * **Application** is a collection of processes in a job
   * **Job** is a collection of applications
   * **Session** is a collection of jobs.

Reply:
Just FWIW: I'm not a big fan of reinventing the wheel, so I simply looked up the definition of "application" in a few dictionaries:

- Websters: a sequence of coded instructions  (such as a word processor or a spreadsheet) that can be inserted into a mechanism to perform a particular task or set of tasks
- Webopedia: An organized list of instructions that, when executed, causes the computer to behave in a predetermined manner.
- Dictionary.com: a precise sequence of instructions enabling a computer to perform a task;
- Oxford Dictionary of Computer Science: “A particular role or task to which a computer system can be applied, or, more usually, the software used for such a purpose”. “Any program that is specific to a particular role and makes a direct contribution to performing that role. For example, where a computer handles a company’s finances a payroll program would be an applications program. By contrast, an *operating system or a *software tool may both be essential to the effective use of the computer system, but neither makes a direct contribution to meeting the end-user’s eventual needs.”

> Process is smallest unit managed by PMIx

Please note that PMIx does not _manage_ anything. Other than that, your hierarchy looks correct to me.

Reply:
So, whether PMIx manages anything is something we were pondering in our last meeting.   The Job Control related functions do some things that look a little bit like managing and I wonder if we need to be clear about how the "things" controlled by Job Control map to the PMix clients (or the "things" that were created by PMIx, but have not called PMIx_Init).    You can pause, resume, kill, signal, terminate based on a namespace/rank specification.  What does that namespace/rank designate.  Is it only the client that called PMIx_Init or is it the specific thing that PMIx started?  I think some of this will have to be system/implementation specific, but we may want to say something about the distinction.